// @prismicio/mocks/lib/renderers/MockContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ctx = {
    urlRewriter: {
        optimizeImageUrl(originUrl) {
            return originUrl;
        },
        rewriteImageUrl(view) {
            return view.url || "/mockURL";
        },
        rewriteFileUrl(originUrl) {
            return originUrl;
        },
        rewriteS3Bucket(url) {
            return url;
        },
        enforceCDN(url) {
            return url;
        },
    },
    emptyStringInsteadOfNull: false,
    Extension: {
        DocEncoder: {
            encodeDocId: (documentId, _fetchUuidOrRelationName, additionalData) => {
                return `${documentId}${(additionalData === null || additionalData === void 0 ? void 0 : additionalData.length) ? "%%" + additionalData.join("%%") : ""}`;
            },
        },
        encoders: {
            documentLinks: IDEncoder("=&="),
            documentUrls: IDEncoder(">&<"),
            integrationFields: IDEncoder("%!%"),
        },
    },
    LinkResolver: {
        buildUrl(_params) {
            return "/";
        },
    },
};
exports.default = ctx;
function IDEncoder(prefix) {
    return {
        encode: (strId) => `${prefix}#${strId}`,
    };
}
