// @prismicio/svelte/dist/kit/redirectToPreviewURL.js
import * as prismic from "@prismicio/client";
/**
 * Redirects a visitor to the URL of a previewed Prismic document from within a
 * SvelteKit `+server` file.
 *
 * Return the created `Response` in your `+server` file.
 *
 * @example
 *
 * ```typescript
 * import { createClient } from "../lib/prismicio.js";
 * import { redirectToPreviewURL } from "@prismicio/svelte/kit";
 *
 * export async function GET({ fetch, request }) {
 * 	const client = createClient({ fetch });
 *
 * 	return await redirectToPreviewURL({ client, request });
 * }
 * ```
 */
export const redirectToPreviewURL = async (config) => {
    const previewToken = new URL(config.request.url).searchParams.get("token") ?? undefined;
    const documentID = new URL(config.request.url).searchParams.get("documentId") ?? undefined;
    const routePrefix = config.routePrefix ?? "preview";
    const previewURL = await config.client.resolvePreviewURL({
        previewToken,
        documentID,
        defaultURL: config.defaultURL || "/",
    });
    // Prevent a flash of non-preview content by setting the preview token
    // on the initial page load.
    if (previewToken) {
        config.cookies.set(prismic.cookie.preview, previewToken, {
            path: "/",
            httpOnly: false,
        });
    }
    return new Response(undefined, {
        status: 307,
        headers: {
            Location: "/" + routePrefix + previewURL,
        },
    });
};
