// @prismicio/custom-types-client/dist/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const client = require("./client.cjs");
const bulkUpdate = require("./bulkUpdate.cjs");
const errors = require("./errors.cjs");
exports.CustomTypesClient = client.CustomTypesClient;
exports.createClient = client.createClient;
exports.BulkUpdateOperationType = bulkUpdate.BulkUpdateOperationType;
exports.BulkUpdateTransaction = bulkUpdate.BulkUpdateTransaction;
exports.createBulkUpdateTransaction = bulkUpdate.createBulkUpdateTransaction;
exports.BulkUpdateHasExistingDocumentsError = errors.BulkUpdateHasExistingDocumentsError;
exports.ConflictError = errors.ConflictError;
exports.ForbiddenError = errors.ForbiddenError;
exports.InvalidAPIResponse = errors.InvalidAPIResponse;
exports.InvalidPayloadError = errors.InvalidPayloadError;
exports.MissingFetchError = errors.MissingFetchError;
exports.NotFoundError = errors.NotFoundError;
exports.PrismicError = errors.PrismicError;
exports.UnauthorizedError = errors.UnauthorizedError;
