// @slicemachine/adapter-sveltekit/dist/hooks/snippet-read.js
import { stripIndent } from "common-tags";
const dotPath = (...segments) => {
  return segments.flat().join(".");
};
const format = async (input, helpers) => {
  const formattedInput = await helpers.format(input, void 0, {
    includeNewlineAtEnd: false,
    prettier: {
      plugins: ["prettier-plugin-svelte"],
      parser: "svelte"
    }
  });
  return formattedInput;
};
const snippetRead = async (data, { helpers }) => {
  var _a, _b;
  const { fieldPath, itemName } = data;
  const label = "Svelte";
  switch (data.model.type) {
    case "StructuredText": {
      return [
        {
          label: `${label} (components)`,
          language: "svelte",
          code: await format(stripIndent`
							<PrismicRichText field={${dotPath(fieldPath)}} />
						`, helpers)
        },
        {
          label: `${label} (plain text)`,
          language: "svelte",
          code: await format(stripIndent`
							<PrismicText field={${dotPath(fieldPath)}} />
						`, helpers)
        }
      ];
    }
    case "Link": {
      const repeat = ((_a = data.model.config) == null ? void 0 : _a.repeat) ?? false;
      const allowText = ((_b = data.model.config) == null ? void 0 : _b.allowText) ?? false;
      let codeText;
      if (!repeat && !allowText) {
        codeText = stripIndent`
					<PrismicLink field={${dotPath(fieldPath)}}>Link</PrismicLink>
				`;
      } else if (!repeat && allowText) {
        codeText = stripIndent`
					<PrismicLink field={${dotPath(fieldPath)}} />
				`;
      } else if (repeat && !allowText) {
        codeText = stripIndent`
					{#each ${dotPath(fieldPath)} as link (link.key)}
						<PrismicLink field={link}>Link</PrismicLink>
					{/each}
				`;
      } else if (repeat && allowText) {
        codeText = stripIndent`
					{#each ${dotPath(fieldPath)} as link (link.key)}
						<PrismicLink field={link} />
					{/each}
				`;
      } else {
        throw new Error("Invalid configuration.");
      }
      return {
        label,
        language: "svelte",
        code: await format(codeText, helpers)
      };
    }
    case "Image": {
      return {
        label,
        language: "svelte",
        code: await format(stripIndent`
							<PrismicImage field={${dotPath(fieldPath)}} />
						`, helpers)
      };
    }
    case "Group": {
      const code = await format(stripIndent`
					{#each ${dotPath(fieldPath)} as ${itemName}}
						<!-- Render content for ${itemName} -->
					{/each}
				`, helpers);
      return {
        label,
        language: "svelte",
        code
      };
    }
    case "Slices": {
      const code = await format(stripIndent`
					<SliceZone
					  slices={${dotPath(fieldPath)}}
					  {components}
					/>
				`, helpers);
      return {
        label,
        language: "svelte",
        code
      };
    }
    case "GeoPoint": {
      const code = await format(stripIndent`
					{${dotPath(fieldPath, "latitude")}}, {${dotPath(fieldPath, "longitude")}}
				`, helpers);
      return {
        label,
        language: "svelte",
        code
      };
    }
    case "Embed": {
      const code = await format(stripIndent`
					{@html ${dotPath(fieldPath, "html")}}
				`, helpers);
      return {
        label,
        language: "svelte",
        code
      };
    }
    default: {
      return {
        label,
        language: "svelte",
        code: await format(stripIndent`
						{${dotPath(fieldPath)}}
					`, helpers)
      };
    }
  }
};
export {
  snippetRead
};
